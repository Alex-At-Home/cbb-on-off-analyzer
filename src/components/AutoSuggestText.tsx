// React imports:
import React, { useState, useEffect, createRef } from 'react';
import Router from 'next/router';

// Lodash
import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';

// Library imports:
import fetch from 'isomorphic-unfetch';

// Additional components:
// @ts-ignore
import TextInput from 'react-autocomplete-input';
import 'react-autocomplete-input/dist/bundle.css';

// Utils:
import { ParamPrefixes, ParamDefaults, GameFilterParams } from '../utils/FilterModels';
import { RosterCompareModel } from '../components/RosterCompareTable';
import { dataLastUpdated } from '../utils/internal-data/dataLastUpdated';
import { ClientRequestCache } from '../utils/ClientRequestCache';
import { QueryUtils } from "../utils/QueryUtils";

/** The keydown event does not come from AutoSuggestText element */
export const notFromAutoSuggest = (event: any) => {
  return (event?.srcElement?.className?.indexOf("auto-suggest") < 0);
}

type Props = {
  readOnly: boolean,
  placeholder: string,
  initValue: string,
  team?: string,
  gender?: string,
  year?: string,
  onChange: (ev: any) => void,
  onKeyUp: (ev: any) => void
  onKeyDown: (ev: any) => void
};
const AutoSuggestText: React.FunctionComponent<Props> = (
  {readOnly, placeholder, initValue, team, year, gender, onChange, onKeyUp, onKeyDown}
) => {

  // Data model

  const [ savedParams, setSavedParams ] = useState({} as GameFilterParams);
  const [ basicOptions, setBasicOptions ] = useState([] as Array<string>);
  const [ advOptions, setAdvOptions ] = useState([] as Array<string>);

  const isDebug = (process.env.NODE_ENV !== 'production');

  const basicOperators = [
    "AND", "OR", "NOT"
  ];
  const advancedFields = basicOperators.concat([
    "players.id:", "opponent.team:", "start_min:", "end_min:",
    "location_type:", "location_type:Home", "location_type:Away", "location_type:Neutral",
    "date:",
    "players_in.id:",  "players_out.id:",
    "score_info.start_diff:", "score_info.end_diff:"
  ]);

  const textRef = createRef();

  // Utils

  /** Reset everything if team/year/gender changes */
  useEffect(() => {
    const params: GameFilterParams = { year: year, gender: gender, team: team }
    if (!_.isEqual(params, savedParams)) {
      if (isDebug) console.log(`Update params: old=[${JSON.stringify(savedParams)}] vs new=[${JSON.stringify(params)}]`);
      setSavedParams(params);
      setBasicOptions([]);
      setAdvOptions([]);
    }
  });

  /** Makes an API call to elasticsearch to get the roster */
  const fetchRoster = () => {
    if (gender && year && team) {
      const genderYear = `${gender}_${year}`;
      const currentJsonEpoch = dataLastUpdated[genderYear] || -1;

      const query: GameFilterParams = {
        gender: gender, year: year, team: team,
        baseQuery: "", onQuery: "", offQuery: "",
        minRank: ParamDefaults.defaultMinRank, maxRank: ParamDefaults.defaultMaxRank
      };
      const paramStr = QueryUtils.stringify(query);
      // Check if it's in the cache:
      const cachedJson = ClientRequestCache.decacheResponse(
        paramStr, ParamPrefixes.roster, currentJsonEpoch, isDebug
      );
      if (cachedJson && !_.isEmpty(cachedJson)) { //(ignore placeholders here)
        handleResponse(cachedJson);
      } else {
        fetch(`/api/getRoster?${paramStr}`).then(function(response: fetch.IsomorphicResponse) {
          response.json().then(function(json: any) {
            // Cache result locally:
            if (isDebug) {
              console.log(`CACHE_KEY=[${ParamPrefixes.roster}${paramStr}]`);
              //(this is a bit chatty)
              //console.log(`CACHE_VAL=[${JSON.stringify(json)}]`);
            }
            if (response.ok) { //(never cache errors)
              ClientRequestCache.cacheResponse(
                paramStr, ParamPrefixes.roster, json, currentJsonEpoch, isDebug
              );
            }
            handleResponse(json);
          });
        })
      }
    }
  };

  /** Parse the return from fetch Roster into name fragments */
  const handleResponse = (json: any) => {
    const jsons = json?.responses || [];
    const rosterCompareJson = (jsons.length > 0) ? jsons[0] : {};
    const roster = rosterCompareJson?.aggregations?.tri_filter?.buckets?.baseline?.player?.buckets || [];

    const nameFrags = _.chain(roster)
      .flatMap((rosterObj) => {
        const nameFragments = _.split(rosterObj.key, /\s*,\s*|\s+/);
        return _.chain(nameFragments)
          .filter((s) => s.length >= 2)
          .value();
      }).sortBy().sortedUniq().value();

    const names = _.chain(roster)
      .map((rosterObj) => `"${rosterObj.key}"`)
      .sortBy().sortedUniq().value();

    const namesAndFrags = nameFrags.concat(names); //(get the order to work well in practice)

    setBasicOptions(namesAndFrags.concat(basicOperators));
    setAdvOptions(namesAndFrags.concat(advancedFields));
  };

  // View

  return <TextInput
    ref={textRef}
    Component={"input"}
    defaultValue={initValue}
    readOnly={readOnly}
    className="form-control auto-suggest"
    placeholder={placeholder}
    requestOnlyIfNoOptions={true} //(only requests if empty)
    options={
      (initValue && ('[' == initValue[0])) ? advOptions : basicOptions
    }
    onRequestOptions = {fetchRoster}
    trigger=""
    regex='^[A-Za-z0-9\\-_"]+$'
    matchAny={true}
    maxOptions={18}
    spaceRemovers={[';', ')', ':', ']']}
    onChange={(eventText: string) => onChange({target: { value: eventText}})}
    onBlur={(ev:any) => {
      const currentTextRef = textRef.current as any;
      setTimeout(() => { //(give out of order events a chance!)
        try {
          currentTextRef.resetHelper();
        } catch (e) {}
      }, 100);
    }}
    onKeyUp={onKeyUp}
    onKeyDown={(ev:any) => {
      // Understanding this requires understanding of internals:
      //https://github.com/yury-dymov/react-autocomplete-input/blob/master/src/AutoCompleteTextField.js
      if (ev.keyCode == 9) {
        const underlyingObj = textRef.current as any;
        if (underlyingObj.state.helperVisible) {
          ev.preventDefault();
          ev.keyCode = 13;
          (textRef.current as any).handleKeyDown(ev);
        }
        //(else will just get passed up)
      } else {
        //(doesn't work for enter/return because of the CommonFilter-specific handler)
        onKeyDown(ev);
      }
    }}
  />;
}
export default AutoSuggestText;
