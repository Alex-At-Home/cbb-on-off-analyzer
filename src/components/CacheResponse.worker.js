import { ClientRequestCache } from '../utils/ClientRequestCache';

console.log("Starting CacheResponseWorker");

self.addEventListener("message", (event) => {
  /**/
  console.log("?GET MESSAGE " + JSON.stringify(event.data));
});
