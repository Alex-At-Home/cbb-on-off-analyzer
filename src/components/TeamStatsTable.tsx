// React imports:
import React from 'react';

// Next imports:
import { NextPage } from 'next';

// Lodash:
import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

// Additional components:
// @ts-ignore
import LoadingOverlay from 'react-loading-overlay';

// Component imports
import GenericTable, { GenericTableOps, GenericTableColProps } from "./GenericTable"

// Util imports
import { CbbColors } from "../utils/CbbColors"
import { CommonTableDefs } from "../utils/CommonTableDefs"

export type TeamStatsModel = {
  on: any,
  off: any,
  baseline: any,
  error_code?: string
}
type Props = {
  teamStats: TeamStatsModel
}

const TeamStatsTable: React.FunctionComponent<Props> = ({teamStats}) => {

  const offPrefixFn = (key: string) => "off_" + key;
  const offCellMetaFn = (key: string, val: any) => "off";
  const defPrefixFn = (key: string) => "def_" + key;
  const defCellMetaFn = (key: string, val: any) => "def";

  const teamStatsOn = { off_title:  "'On' Offense", def_title: "'On' Defense", ...teamStats.on };
  const teamStatsOff = { off_title:"'Off' Offense", def_title: "'Off' Defense", ...teamStats.off };
  const teamStatsBaseline = { off_title: "'Baseline' Offense", def_title: "'Baseline' Defense", ...teamStats.baseline };

  const tableData = _.flatMap([
    (teamStats.on?.doc_count) ? [
      GenericTableOps.buildDataRow(teamStatsOn, offPrefixFn, offCellMetaFn),
      GenericTableOps.buildDataRow(teamStatsOn, defPrefixFn, defCellMetaFn),
      GenericTableOps.buildRowSeparator()
    ] : [],
    (teamStats.off?.doc_count) ? [
      GenericTableOps.buildDataRow(teamStatsOff, offPrefixFn, offCellMetaFn),
      GenericTableOps.buildDataRow(teamStatsOff, defPrefixFn, defCellMetaFn),
      GenericTableOps.buildRowSeparator()
    ] : [],
    [ GenericTableOps.buildDataRow(teamStatsBaseline, offPrefixFn, offCellMetaFn),
    GenericTableOps.buildDataRow(teamStatsBaseline, defPrefixFn, defCellMetaFn),
    GenericTableOps.buildRowSeparator() ],
  ]);

  // 2] Data Model

  // 3] Utils
  /** Sticks an overlay on top of the table if no query has ever been loaded */
  function needToLoadQuery() {
    return (Object.keys(teamStats.on).length == 0) &&
      (Object.keys(teamStats.off).length == 0) &&
      (Object.keys(teamStats.baseline).length == 0);
  }

  // 4] View
  return <Container>
    <LoadingOverlay
      active={needToLoadQuery()}
      text={teamStats.error_code ?
        `Query Error: ${teamStats.error_code}` :
        "Press 'Submit' to view results"
      }
    >
      <Row>
        <Col>
          <GenericTable tableCopyId="teamStatsTable" tableFields={CommonTableDefs.onOffTable} tableData={tableData}/>
        </Col>
      </Row>
    </LoadingOverlay>
  </Container>;
};

export default TeamStatsTable;
