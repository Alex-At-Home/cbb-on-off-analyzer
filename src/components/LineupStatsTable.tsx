// React imports:
import React, { useState, useEffect } from 'react';

// Next imports:
import { NextPage } from 'next';

// Lodash:
import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import InputGroup from 'react-bootstrap/InputGroup';

// Additional components:
// @ts-ignore
import LoadingOverlay from 'react-loading-overlay';
import Select, { components} from "react-select";

// Component imports
import GenericTable, { GenericTableOps, GenericTableColProps } from "./GenericTable";
import { LineupFilterParams, ParamDefaults } from '../utils/FilterModels';
import { RosterStatsModel } from '../components/RosterStatsTable';

// Util imports
import { CbbColors } from "../utils/CbbColors";
import { CommonTableDefs } from "../utils/CommonTableDefs";
import { PositionUtils } from "../utils/stats/PositionUtils";

export type LineupStatsModel = {
  lineups?: Array<any>,
  error_code?: string
}
type Props = {
  lineupStats: LineupStatsModel,
  rosterStats: RosterStatsModel,
  startingState: LineupFilterParams;
  onChangeState: (newParams: LineupFilterParams) => void;
}

const LineupStatsTable: React.FunctionComponent<Props> = ({lineupStats, rosterStats, startingState, onChangeState}) => {

  // 1] State

  const [ minPoss, setMinPoss ] = useState(startingState.minPoss || ParamDefaults.defaultLineupMinPos);
  const [ maxTableSize, setMaxTableSize ] = useState(startingState.maxTableSize || ParamDefaults.defaultLineupMaxTableSize);
  const [ sortBy, setSortBy ] = useState(startingState.sortBy || ParamDefaults.defaultLineupSortBy);
  const [ filterStr, setFilterStr ] = useState(startingState.filter || ParamDefaults.defaultLineupFilter);

  // (slight delay when typing into the filter to make it more responsive)
  const [ timeoutId, setTimeoutId ] = useState(-1);
  const [ tmpFilterStr, setTmpFilterStr ] = useState(filterStr);

  const filterFragments =
    filterStr.split(",").map(fragment => _.trim(fragment)).filter(fragment => fragment ? true : false);
  const filterFragmentsPve =
    filterFragments.filter(fragment => fragment[0] != '-');
  const filterFragmentsNve =
    filterFragments.filter(fragment => fragment[0] == '-').map(fragment => fragment.substring(1));

  useEffect(() => { //(this ensures that the filter component is up to date with the union of these fields)
    const newState = _.merge(startingState, {
      minPoss: minPoss,
      maxTableSize: maxTableSize,
      sortBy: sortBy,
      filter: filterStr
    });
    onChangeState(newState);
  }, [ minPoss, maxTableSize, sortBy, filterStr ]);

  // 2] Data Model

  // 3] Utils

  // 3.1] Build individual info

  // 3.1.1] Positional info from the season stats

  const positionFromPlayerKey = _.chain(rosterStats.global || []).map((player: any) => {
    const [ posConfs, posConfsDiags ] = PositionUtils.buildPositionConfidences(player);
    const [ pos, posDiags ] = PositionUtils.buildPosition(posConfs, player);
    return [ player.key, { posConfidences: _.values(posConfs || {}), posClass: pos } ];
  }).fromPairs().value();

  // 3.2] Table building

  const offPrefixFn = (key: string) => "off_" + key;
  const offCellMetaFn = (key: string, val: any) => "off";
  const defPrefixFn = (key: string) => "def_" + key;
  const defCellMetaFn = (key: string, val: any) => "def";

  const sorter = (sortStr: string) => { // format: (asc|desc):(off_|def_|diff_)<field>
    const sortComps = sortStr.split(":"); //asc/desc
    const dir = (sortComps[0] == "desc") ? -1 : 1;
    const fieldComps = _.split(sortComps[1], "_", 1); //off/def/diff
    const fieldName = sortComps[1].substring(fieldComps[0].length + 1); //+1 for _
    const field = (lineup: any) => {
      switch(fieldComps[0]) {
        case "diff": //(off-def)
          return (lineup["off_" + fieldName]?.value || 0.0)
                - (lineup["def_" + fieldName]?.value || 0.0);
        default: return lineup[sortComps[1]]?.value; //(off or def)
      }
    };
    return (lineup: any) => {
      return dir*(field(lineup) || 0);
    };
  };

  const lineups = lineupStats?.lineups || [];
  const tableData = _.chain(lineups).filter((lineup) => {
      const minPossInt = parseInt(minPoss);
      const offPos = lineup.off_poss?.value || 0;
      const defPos = lineup.def_poss?.value || 0;
      return offPos >= minPossInt || defPos >= minPossInt; //(unclear which of || vs && is best...)
    }).filter((lineup) => {
      const namesToTest = _.chain(lineup?.players_array?.hits?.hits).flatMap((pDoc) => {
        return pDoc?._source.players.map((p: any) => p.id);
      }).value().concat([ lineup.key ]);

      const playerFilter =
        (_.isEmpty(filterFragmentsPve) ||
          _.every(filterFragmentsPve, (frag) => _.some(namesToTest, (name) => name.indexOf(frag) >= 0))
        ) &&
        (_.isEmpty(filterFragmentsNve) ||
          !_.some(filterFragmentsNve, (frag) => _.some(namesToTest, (name) => name.indexOf(frag) >= 0))
        );
      return playerFilter && (lineup.key != ""); // (workaround for #53 pending fix)

    }).sortBy(
       [ sorter(sortBy) ]
    ).take(
      parseInt(maxTableSize)
    ).flatMap((lineup) => {
      const codesAndIds = lineup.players_array?.hits?.hits?.[0]?._source?.players || [];

      const sortedCodesAndIds = PositionUtils.orderLineup(codesAndIds, positionFromPlayerKey);

      const title = sortedCodesAndIds.map((cid: { code: string, id: string}) => cid.code).join(" / ");
      const stats = { off_title: title, def_title: "", ...lineup };
      return [
        GenericTableOps.buildDataRow(stats, offPrefixFn, offCellMetaFn),
        GenericTableOps.buildDataRow(stats, defPrefixFn, defCellMetaFn),
        GenericTableOps.buildRowSeparator()
      ];
    }).value();

  // 3.2] Sorting utils

  const sortOptions: Array<any> = _.flatten(
    _.toPairs(CommonTableDefs.lineupTable)
      .filter(keycol => keycol[1].colName && keycol[1].colName != "")
      .map(keycol => {
        return [
          ["desc","off"], ["asc","off"], ["desc","def"], ["asc","def"], ["desc","diff"], ["asc","diff"]
        ].map(combo => {
          const ascOrDesc = (s: string) => { switch(s) {
            case "asc": return "Asc.";
            case "desc": return "Desc.";
          }}
          const offOrDef = (s: string) => { switch(s) {
            case "off": return "Offensive";
            case "def": return "Defensive";
            case "diff": return "Off-Def";
          }}
          return {
            label: `${keycol[1].colName} (${ascOrDesc(combo[0])} / ${offOrDef(combo[1])})`,
            value: `${combo[0]}:${combo[1]}_${keycol[0]}`
          };
        });
      })
  );
  const sortOptionsByValue = _.fromPairs(
    sortOptions.map(opt => [opt.value, opt])
  );
  /** Put these options at the front */
  const mostUsefulSubset = [
    "desc:off_poss",
    "desc:diff_adj_ppp",
    "desc:off_adj_ppp",
    "asc:def_adj_ppp",
  ];
  /** The two sub-headers for the dropdown */
  const groupedOptions = [
    {
      label: "Most useful",
      options: _.chain(sortOptionsByValue).pick(mostUsefulSubset).values().value()
    },
    {
      label: "Other",
      options: _.chain(sortOptionsByValue).omit(mostUsefulSubset).values().value()
    }
  ];
  /** The sub-header builder */
  const formatGroupLabel = (data: any) => (
    <div>
      <span>{data.label}</span>
    </div>
  );

  // 3] Utils
  function picker(offScale: (val: number) => string, defScale: (val: number) => string) {
    return (val: any, valMeta: string) => {
      const num = val.value as number;
      return _.isNil(num) ?
        CbbColors.malformedDataColor : //(we'll use this color to indicate malformed data)
        ("off" == valMeta) ? offScale(num) : defScale(num)
        ;
    };
  }
  /** Sticks an overlay on top of the table if no query has ever been loaded */
  function needToLoadQuery() {
    return lineupStats.lineups === undefined;
  }
  function rowSpanCalculator(cellMeta: string) {
    switch(cellMeta) {
      case "off": return 2;
      case "def": return 0;
      default: return 1;
    }
  }

  /** For use in selects */
  function stringToOption(s: string) {
    return sortOptionsByValue[s];
  }

  /** Handling filter change (/key presses to fix the select/delete on page load) */
  const onFilterChange = (ev: any) => {
    const toSet = ev.target.value;
    setTmpFilterStr(toSet);
    if (timeoutId != -1) {
      window.clearTimeout(timeoutId);
    }
    setTimeoutId(window.setTimeout(() => {
      setFilterStr(toSet);
    }, 100));
  };

  // 4] View
  return <Container>
    <LoadingOverlay
      active={needToLoadQuery()}
      text={lineupStats.error_code ?
        `Query Error: ${lineupStats.error_code}` :
        "Press 'Submit' to view results"
      }
    >
      <Form.Row>
        <Form.Group as={Col} sm="6">
          <InputGroup>
            <InputGroup.Prepend>
              <InputGroup.Text id="filter">Filter</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control
              onKeyUp={onFilterChange}
              onChange={onFilterChange}
              placeholder = "eg Player1Surname,Player2FirstName,-Player3Name"
              value={tmpFilterStr}
            />
          </InputGroup>
        </Form.Group>
      </Form.Row>
      <Form.Row>
        <Form.Group as={Col} sm="3">
          <InputGroup>
            <InputGroup.Prepend>
              <InputGroup.Text id="maxLineups">Max Lineups</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control
              onChange={(ev: any) => {
                if (ev.target.value.match("^[0-9]*$") != null) {
                  setMaxTableSize(ev.target.value);
                }
              }}
              placeholder = "eg 50"
              value={maxTableSize}
            />
          </InputGroup>
        </Form.Group>
        <Form.Group as={Col} sm="3">
          <InputGroup>
            <InputGroup.Prepend>
              <InputGroup.Text id="minPossessions">Min Poss #</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control
              onChange={(ev: any) => {
                if (ev.target.value.match("^[0-9]*$") != null) {
                  setMinPoss(ev.target.value);
                }
              }}
              placeholder = "eg 5"
              value={minPoss}
            />
          </InputGroup>
        </Form.Group>
        <Form.Group as={Col} sm="6">
          <InputGroup>
            <InputGroup.Prepend>
              <InputGroup.Text id="sortBy">Sort By</InputGroup.Text>
            </InputGroup.Prepend>
            <Select
              className="w-75"
              value={ stringToOption(sortBy) }
              options={ groupedOptions }
              onChange={(option) => { if ((option as any)?.value)
                setSortBy((option as any)?.value);
              }}
              formatGroupLabel={formatGroupLabel}
            />
          </InputGroup>
        </Form.Group>
      </Form.Row>
      <Row>
        <Col>
          <GenericTable tableCopyId="lineupStatsTable" tableFields={CommonTableDefs.lineupTable} tableData={tableData}/>
        </Col>
      </Row>
    </LoadingOverlay>
  </Container>;
};

export default LineupStatsTable;
