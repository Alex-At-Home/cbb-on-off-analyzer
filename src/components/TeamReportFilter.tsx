// React imports:
import React, { useState, useEffect } from 'react';

// Next imports:
import { NextPage } from 'next';

// Lodash:
import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

// Component imports:
import { efficiencyAverages } from '../utils/public-data/efficiencyAverages';
import { LineupStatsModel } from '../components/LineupStatsTable';
import CommonFilter, { CommonFilterParams } from '../components/CommonFilter';
import { ParamPrefixes, TeamReportFilterParams } from "../utils/FilterModels";

// Library imports:
import fetch from 'isomorphic-unfetch';

type Props = {
  onStats: (reportStats: LineupStatsModel) => void;
  startingState: TeamReportFilterParams;
  onChangeState: (newParams: TeamReportFilterParams) => void;
}

const TeamReportFilter: React.FunctionComponent<Props> = ({onStats, startingState, onChangeState}) => {

  // Data model

  /** The state managed by the CommonFilter element */
  const [ commonParams, setCommonParams ] = useState({
      year: startingState.year, team: startingState.team, gender: startingState.gender,
      minRank: startingState.minRank, maxRank: startingState.maxRank,
  } as CommonFilterParams);

  // Lineup Filter - custom queries and filters:

  const [ baseQuery, setBaseQuery ] = useState(startingState.lineupQuery || "");

  const isDebug = (process.env.NODE_ENV !== 'production');

  const cacheKeyPrefix = ParamPrefixes.report;

  // Utils

  /** Bridge between the callback in CommonFilter and state management */
  function updateCommonParams(params: CommonFilterParams) {
    setCommonParams(params)
  }

  /** Builds a game filter from the various state elements */
  function buildParamsFromState(includeFilterParams: Boolean): TeamReportFilterParams {
    return includeFilterParams ?
    _.merge(
      buildParamsFromState(false), {
        sortBy: startingState.sortBy,
        filter: startingState.filter,
        showComps: startingState.showComps,
    }) : {
      team: commonParams.team,
      year: commonParams.year,
      gender: commonParams.gender,
      lineupQuery: baseQuery,
      minRank: commonParams.minRank,
      maxRank: commonParams.maxRank
    };
  }

  /** Handles the response from ES to a stats calc request */
  function handleResponse(json: any, wasError: Boolean) {
    const jsons = json?.responses || [];
    const lineupJson = (jsons.length > 0) ? jsons[0] : {};
    onStats({
      lineups: lineupJson?.aggregations?.lineups?.buckets,
      error_code: wasError ? (lineupJson?.status || json?.status) : undefined,
      avgOff: efficiencyAverages[`${commonParams.gender}_${commonParams.year}`]
    });
  }

  /** Builds the query issued to the API server - the response handling is generic */
  function onSubmit(paramStr: string, callback: (resp: fetch.IsomorphicResponse) => void) {
    fetch(`/api/calculateLineupStats?${paramStr}`).then(callback);
  }

  // Visual components:

  /** Let the user know that he might need to change */

  return <CommonFilter //(generic type inferred)
      startingState={startingState}
      onChangeState={onChangeState}
      onChangeCommonState={updateCommonParams}
      tablePrefix = {cacheKeyPrefix}
      buildParamsFromState={buildParamsFromState}
      childHandleResponse={handleResponse}
      childSubmitRequest={onSubmit}
    >
      <Form.Group as={Row}>
        <Form.Label column sm="2">Baseline Query <a href="https://hoop-explorer.blogspot.com/2020/01/basic-and-advanced-queries-in-hoop.html" target="_blank">(?)</a></Form.Label>
        <Col sm="8">
          <Form.Control
            placeholder="eg 'Player1 AND NOT (WalkOn1 OR WalkOn2)'"
            value={baseQuery}
            onKeyUp={(ev: any) => setBaseQuery(ev.target.value)}
            onChange={(ev: any) => setBaseQuery(ev.target.value)}
          />
        </Col>
        <Col sm="2">
          <Form.Check type="switch"
            id="excludeWalkons"
            checked={false}
            disabled
            label="Auto Walk-ons"
          />
        </Col>
      </Form.Group>
    </CommonFilter>
    ;
}

export default TeamReportFilter;
