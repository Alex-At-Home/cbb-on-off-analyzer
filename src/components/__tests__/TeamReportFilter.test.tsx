import React from 'react';
import TeamReportFilter from '../TeamReportFilter';
import { TeamReportFilterParams } from "../utils/FilterModels";
import { LineupStatsModel } from '../components/LineupStatsTable';
import { RosterStatsModel } from '../components/RosterStatsTable';
import { TeamStatsModel } from '../components/TeamStatsTable';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe("TeamReportFilter", () => {
  test("TeamReportFilter - should create snapshot", () => {
    const dummySubmitCallback = (lineupStats: LineupStatsModel, teamStats: TeamStatsModel, rosterStats: RosterStatsModel) => {};
    const dummyChangeStateCallback = (stats: TeamReportFilterParams) => {};
    const wrapper = shallow(<TeamReportFilter
      onStats={dummySubmitCallback}
      startingState={{}}
      onChangeState={dummyChangeStateCallback}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
//TODO: add mock for fetch and test submit etc
//TODO: check text
