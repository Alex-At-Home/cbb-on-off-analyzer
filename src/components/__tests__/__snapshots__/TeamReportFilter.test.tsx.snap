// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamReportFilter TeamReportFilter - should create snapshot 1`] = `
<CommonFilter
  buildParamsFromState={[Function]}
  childHandleResponse={[Function]}
  onChangeCommonState={[Function]}
  onChangeState={[Function]}
  startingState={Object {}}
  tablePrefix="report-"
/>
`;
